public/app/features/alerting/unified/triage/scene/dataTransform.ts

```diff
@@ -0,0 +1,158 @@
+import { DataFrame } from '@grafana/data';
+
+import { AlertRuleRow, EmptyLabelValue, GenericGroupedRow, WorkbenchRow } from '../types';
+
+// Single-pass implementation: 30x faster than previous approach
+// Builds tree structure in one pass through data, avoiding intermediate row objects
+export function convertToWorkbenchRows(series: DataFrame[], groupBy: string[] = []): WorkbenchRow[] {
```

Comment:
This algorithm seem to be faster. I run a few benchmarks to compare its performance with the old one and I feel the results are worth the complexity

| Test | Old (ops/sec) | Old (avg time) | New (ops/sec) | New (avg time) | Speedup |
|------|---------------|----------------|---------------|----------------|---------|
| No grouping | 11.26 | 88.790 ms | 384.62 | 2.600 ms | 34.2x |
| Single groupBy (grafana_folder) | 6.55 | 152.710 ms | 59.31 | 16.860 ms | 9.1x |
| Double groupBy (grafana_folder, alertstate) | 6.13 | 163.020 ms | 43.55 | 22.960 ms | 7.1x |